<a href="https://colab.research.google.com/github/rxphaelbihag/DS171-Intro-to-Supervised-Machine-Learning/blob/main/LaboratoryExercise3_BayesianRegression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Laboratory Exercise 3: Bayesian Regression

Name: Raphael Khandie B. Bihag

Student No.: 2024-05203

## The Three-Card Problem

The three-card problem is a classic probability puzzle involving three cards: one blue/blue, one blue/red, and one red/red. If a card is chosen at random and you see a blue side, the probability that the other side is also blue is 2/3, not 1/2. This happens because the observation "blue side" can only come from two cards, but the blue/blue card is twice as likely to produce a blue side than the mixed blue/red card.

In [3]:
def infer_posterior(observed_color="red"):
    cards = [
        ["blue", "blue"],
        ["blue", "red"],
        ["red", "red"]
    ]

    priors = [1, 1, 1]

    # stores the posteriors
    posterior_unnormalized = {}

    for card, prior in zip(cards, priors):

        # Calculates the likelihood of drawing the observed color from this card
        p_color_given_card = card.count(observed_color) / len(card)

        # Unnormalized posterior: Prior * Likelihood
        joint_prob = prior * p_color_given_card

        card_key = "-".join(card)
        posterior_unnormalized[card_key] = joint_prob

    # marginal likelihood of observing the target color
    total_weight = sum(posterior_unnormalized.values())

    # Handle edge case where total_weight is 0
    if total_weight == 0:
        return {card_key: 0.0 for card_key in posterior_unnormalized}

    # Normalize probabilities
    return {card: prob / total_weight for card, prob in posterior_unnormalized.items()}

## Evaluation

In [4]:
print("Posterior given observed face is BLUE")
for card, prob in infer_posterior(observed_color="blue").items():
    print(f"{card}: {prob:.4f}")

Posterior given observed face is BLUE
blue-blue: 0.6667
blue-red: 0.3333
red-red: 0.0000


In [5]:
print("Posterior given observed face is RED")
for card, prob in infer_posterior(observed_color="red").items():
    print(f"{card}: {prob:.4f}")

Posterior given observed face is RED
blue-blue: 0.0000
blue-red: 0.3333
red-red: 0.6667


## Version with Different Priors
This version has the priors updated such that Card 1 (RR) is twice as likely to be drawn as Card 2 (BB) or Card 3 (RB).

In [6]:
def infer_posterior(observed_color="red"):
    cards = [
        ["blue", "blue"],
        ["blue", "red"],
        ["red", "red"]
    ]

    # prior 2 for the red-red card
    priors = [1, 1, 2]

    # stores the posteriors
    posterior_unnormalized = {}

    for card, prior in zip(cards, priors):

        # Calculates the likelihood of drawing the observed color from this card
        p_color_given_card = card.count(observed_color) / len(card)

        # Unnormalized posterior: Prior * Likelihood
        joint_prob = prior * p_color_given_card

        card_key = "-".join(card)
        posterior_unnormalized[card_key] = joint_prob

    # marginal likelihood of observing the target color
    total_weight = sum(posterior_unnormalized.values())

    # Handle edge case where total_weight is 0
    if total_weight == 0:
        return {card_key: 0.0 for card_key in posterior_unnormalized}

    # Normalize probabilities
    return {card: prob / total_weight for card, prob in posterior_unnormalized.items()}


print("Posterior given observed face is RED with updated priors")
for card, prob in infer_posterior(observed_color="red").items():
    print(f"{card}: {prob:.4f}")

Posterior given observed face is RED with updated priors
blue-blue: 0.0000
blue-red: 0.2000
red-red: 0.8000


## Questions

1. Given that the observed side is Red, what is the probability that the other side of the chosen card is also Red? Show both your mathematical calculation (Bayes' Theorem) and your Python output.

> Using Bayes' Theorem, $P(\text{RR} \mid \text{Red}) = \frac{P(\text{Red} \mid \text{RR}) P(\text{RR})}{P(\text{Red})} = \frac{1 \times (1/3)}{1/2} = \frac{2}{3} \approx 66.67\%$, because the total probability of seeing a red face $P(\text{Red})$ across all three equal-weight cards is $\frac{0}{6} + \frac{1}{6} + \frac{2}{6} = \frac{1}{2}$. Running the function at red yields red-red: 0.6667, confirming that there is a 66.67% chance the other side is also red.

2. Explain why the intuitive guess of 50% is incorrect.

> The 50% intuition mistakenly assumes that because the Blue/Blue card is ruled out, we are choosing equally between two remaining physical cards (Red/Red vs. Red/Blue). However, this ignores the individual faces: the Red/Red card contributes two red sides to the overall pool while the Red/Blue card contributes only one. We are twice as likely to have drawn the Red/Red card because it gives us two ways to observe a red face initially.


3. Suppose the hat is loaded such that Card 1 (RR) is twice as likely to be drawn as Card 2 (BB) or Card 3 (RB). Update your Python model with this new prior and compute the updated posterior probability that the other side is Red.

> Assigning relative prior weights of priors = [2, 1, 1] for [["red", "red"], ["blue", "blue"], ["blue", "red"]] changes the unnormalized joint weights for observing a red side to $2 \times 1.0 = 2.0$ for Red/Red, $1 \times 0.0 = 0.0$ for Blue/Blue, and $1 \times 0.5 = 0.5$ for Blue/Red. Normalizing these values ($2.0 / 2.5$) increases the posterior probability of the Red/Red card to $0.80$ or $80\%$. In Python, updating the priors list to [2, 1, 1] outputs red-red: 0.8000.

## References:
* Franke, M. (n.d.). Bayesian Regression: Theory & Practice - Building Bayesian intuitions: Sampling from models of the data-generating process. Github.Io. https://michael-franke.github.io/Bayesian-Regression/practice-sheets/00a-examples-WebPPL.html